### DISCUSSION TITLE
ALT_R fixed on Windows 10 Client (German Keyboard Layout, but might also work for others)

### DISCUSSION THREAD

**Discussion by empusas** [debauchee/barrier#1809] [Show and tell]
So after weeks of frustration I finally got a working solution for the ALT right key on my Germany Keyboard attached to my Raspberry PI Server with Linux and Mac Clients.
My setup:
A raspberry Pi as mouse/keyboard server and Windows and Mac Clients.
On my windows client I could not use the ALT right key, which is required for curly and square brackets, pipe and some other special characters.
I could use Strg+ALT left to emulate the ALT right key, but that was very inconvinitent. 
I suspected this to be an issue with the barrier. I have tried all different workarounds described here, but nothing worked for me. So I bought a license for Synergy in hope of a fix for this problem and I got exactly the same problem. 
OK, the code is nearly identical, the barrier clients even tried to connect to the Synergy server.
But while I was setting up the Syngery clients I noticed something. On a mac keyboard both option keys, left and right have the same function. then I noticed that the right ALT key was not only not working in Windows, but also on Macs and the linux server(Rapsberry Pi) itself.

The root cause was not barrier! This is how I fixed it and explain my way of troubleshooting this as this might also help with other keyboard layouts.

First I figured out my input device. I use a Logitech MX plus keyboard and I use the German layout with no dead keys.

```
pi@raspberrypi:~ $ xinput list
⎡ Virtual core pointer                    	id=2	[master pointer  (3)]
⎜   ↳ Virtual core XTEST pointer              	id=4	[slave  pointer  (2)]
⎜   ↳ Logitech MX Keys                        	id=12	[slave  pointer  (2)]
⎜   ↳ Logitech USB Receiver Mouse             	id=15	[slave  pointer  (2)]
⎜   ↳ Logitech USB Receiver Consumer Control  	id=6	[slave  pointer  (2)]
⎜   ↳ vc4                                     	id=8	[slave  pointer  (2)]
⎜   ↳ vc4                                     	id=9	[slave  pointer  (2)]
⎣ Virtual core keyboard                   	id=3	[master keyboard (2)]
    ↳ Virtual core XTEST keyboard             	id=5	[slave  keyboard (3)]
    ↳ Logitech USB Receiver                   	id=16	[slave  keyboard (3)]
    ↳ Logitech MX Keys                        	id=13	[slave  keyboard (3)]
    ↳ Logitech USB Receiver Consumer Control  	id=7	[slave  keyboard (3)]
    ↳ vc4                                     	id=10	[slave  keyboard (3)]
    ↳ vc4                                     	id=11	[slave  keyboard (3)]
    ↳ Logitech USB Receiver System Control    	id=14	[slave  keyboard (3)]
```

Next I started a test on my keyboard (id 13) and pressed the ALT Right key.

```
pi@raspberrypi:~ $ xinput test 13
key release 36 
key press   108 
key release 108 
key press   37 
key press   54 
^C
```

Then I checked my layout

```
pi@raspberrypi:~ $ setxkbmap -print -query
xkb_keymap {
	xkb_keycodes  { include "evdev+aliases(qwertz)"	};
	xkb_types     { include "complete"	};
	xkb_compat    { include "complete"	};
	xkb_symbols   { include "pc+de(nodeadkeys)+inet(evdev)"	};
	xkb_geometry  { include "pc(pc105)"	};
};
rules:      evdev
model:      pc105
layout:     de
variant:    nodeadkeys
```

and then checked the mapping for key 108 (ALT R)

```
pi@raspberrypi:~ $ xmodmap -pke | grep 108
keycode 108 = ISO_Level3_Shift NoSymbol ISO_Level3_Shift
```

As you can see the ALT_R key (108) on my keyboard was mapped to ISO_Level3_Shift, but not to ALT_R. So I just needed to fix this:

```
pi@raspberrypi:~ $ xmodmap -e "keycode 108 = Alt_R Meta_R Alt_R Meta_R"
pi@raspberrypi:~ $ xmodmap -pke | grep 108
keycode 108 = Alt_R NoSymbol Alt_R
pi@raspberrypi:~ $ 
```

After that change my ALT Right key was working fine on the Windows and Mac clients.

I hope this guide does help you to fix your issues with not working keys as well.

**Comment by MicBoucinha** [comment]
I tried your solution on a 'pt' layout with a Ubuntu server and a Windows client.  It works for the clients but I lost the ability of AltGr on the Ubuntu server. 

Did that also happen to you?

**Reply by empusas** [reply]
Hi, yes that is true. I did not mind because I use the Raspberry PI just as Keybord/Mouse Hub to control other computers and have some tools like PiHole running on it. But I don't really use it for anything else. To me it was just annoying that I could not have any brackets or the pipe sign on my Windows clients.
I guess you could try the key mapping feature in Barrier to map the ISO_Level3_Shift to ALT-R on other computers, but I could not get it to work. And mapping another key to the ALT-R was just as annoying to me as to use STRG + ALT-L all the time.

**Reply by MicBoucinha** [reply]
Thanks, I will try your suggestion.

**Comment by empusas** [comment]
Hi,

this worked for me on my raspberry PI to enable the ALT_R on Windows and have the key still working in Linux on my Pi:

`xmodmap -e "keycode 108 = Alt_R Meta_R Alt_R Meta_R ISO_Level3_Shift"
`
Be also aware that this fix is not permanent. You would have to add it to your .bashrc or something to have it executed every time you login.

Another option might be to create your own "custom" keymap which is documented here:

[Keyboard_configuration](https://wiki.archlinux.org/title/Linux_console/Keyboard_configuration)

**Reply by MicBoucinha** [reply]
Unfortunately that didn't work out in my case. I will keep trying.

Thanks!
